### PR TITLE
FR-5727 - add query keeper feature

### PR DESCRIPTION
**Pull request by MaxArnaut**
Demo 
https://frontegg.atlassian.net/browse/FR-5727

Do not merge need to discuss some points

### CODE REVIEW

File: packages/react/src/queryKeeper.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  useEffect(() => {
+    if (!!history.location.search) {
+      queryParams.current = history.location.search;
+      prevPathname.current = history.location.pathname;
+    }
+  }, []);
+
+  history.listen((listener) => {
+    if (listener.search) {
+      queryParams.current = listener.search;
```
Comment (rotemzif1) [inline packages/react/src/queryKeeper.tsx:33]: what is the difference between the use case in this line to link 26?

File: packages/react/src/queryKeeper.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    const shouldKeepQuery = !!Object.values(routes).find((route) => route === prevPathname.current);
+
+    if (!listener.search && !!queryParams.current && shouldKeepQuery) {
+      const query = removeRedirectUrlFromQuery(queryParams.current);
+      history.push(listener.pathname + `?${query}`);
+      queryParams.current = '';
+      prevPathname.current = '';
```
Comment (rotemzif1) [inline packages/react/src/queryKeeper.tsx:42]: why we need this line if we have line 45?

File: packages/react/src/queryKeeper.tsx (hunk 1 above)
Comment (MaxArnaut) [reply packages/react/src/queryKeeper.tsx]: no need

File: packages/react/src/queryKeeper.tsx (hunk 2 above)
Comment (MaxArnaut) [reply packages/react/src/queryKeeper.tsx]: no need, removed

File: packages/react/src/queryKeeper.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  const history = useHistory();
+
+  useEffect(() => {
+    if (!!history.location.search) {
+      queryParams.current = history.location.search;
+      prevPathname.current = history.location.pathname;
+    }
+  }, []);
+
+  history.listen((listener) => {
```
Comment (frontegg-david) [inline packages/react/src/queryKeeper.tsx:31]: - infinite subscription listener
- callback not in useCallback

File: packages/react/src/FronteggProvider.tsx [hunk 4]
```diff
@@ -65,6 +66,11 @@ export const Connector: FC<ConnectorProps> = ({ history, appName, ...props }) =>
   }, []);
   ContextHolder.setOnRedirectTo(onRedirectTo);
 
+  //If you want to handle exact routes you can pass it like in example below
+  // const routesShouldKeepQuery = {signUpUrl: "/account/sign-up"}
+  // useQueryKeeper({routes: routesShouldKeepQuery});
+  useQueryKeeper({ routes: app.store.getState().auth.routes });
```
Comment (frontegg-david) [inline packages/react/src/FronteggProvider.tsx:72]: This Connector component should not have any rerender mechanisim like useState or useHistory.. and the useQueryKeeper will rerender Connector component for every history change

**Review by frontegg-david — CHANGES_REQUESTED** [review]

**Review by frontegg-david — CHANGES_REQUESTED** [review]

**Review by frontegg-david — APPROVED** [review]

**Pull request merged**